pkg/util/grpc/stats/grpc_stats.go

```diff
@@ -0,0 +1,109 @@
+package stats
+
+import (
+	"context"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"google.golang.org/grpc/stats"
+)
+
+func NewStatsHandler(r prometheus.Registerer) stats.Handler {
+	const MiB = 1024 * 1024
+	messageSizeBuckets := []float64{1 * MiB, 2.5 * MiB, 5 * MiB, 10 * MiB, 25 * MiB, 50 * MiB, 100 * MiB, 250 * MiB}
+
+	return &grpcStatsHandler{
+		connectedClients: promauto.With(r).NewGauge(prometheus.GaugeOpts{
+			Name: "cortex_grpc_connected_clients",
+			Help: "Number of clients connected to gRPC server",
+		}),
+
+		inflightRPC: promauto.With(r).NewGaugeVec(prometheus.GaugeOpts{
+			Name: "cortex_grpc_inflight_requests",
+			Help: "Number of inflight RPC calls",
+		}, []string{"method"}),
+
+		methodErrors: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_grpc_method_errors_total",
+			Help: "Number of clients connected to gRPC server",
+		}, []string{"method"}),
+
+		receivedMessageSize: promauto.With(r).NewHistogramVec(prometheus.HistogramOpts{
+			Name:    "cortex_grpc_request_size_bytes",
+			Help:    "Size of gRPC requests.",
+			Buckets: messageSizeBuckets,
+		}, []string{"method"}),
+
+		sentMessageSize: promauto.With(r).NewHistogramVec(prometheus.HistogramOpts{
+			Name:    "cortex_grpc_response_size_bytes",
+			Help:    "Size of gRPC responses.",
+			Buckets: messageSizeBuckets,
+		}, []string{"method"}),
+	}
+}
+
+type grpcStatsHandler struct {
+	connectedClients    prometheus.Gauge
+	inflightRPC         *prometheus.GaugeVec
+	receivedMessageSize *prometheus.HistogramVec
+	sentMessageSize     *prometheus.HistogramVec
+	methodErrors        *prometheus.CounterVec
+}
+
+// Custom type to hide it from other packages.
+type contextKey int
+
+const (
+	contextKeyMethodName contextKey = 1
+)
+
+func (g *grpcStatsHandler) TagRPC(ctx context.Context, info *stats.RPCTagInfo) context.Context {
+	return context.WithValue(ctx, contextKeyMethodName, info.FullMethodName)
+}
+
+func (g *grpcStatsHandler) HandleRPC(ctx context.Context, rpcStats stats.RPCStats) {
+	// We use full method name from context, because not all RPCStats structs have it.
+	fullMethodName, ok := ctx.Value(contextKeyMethodName).(string)
```

Comment:
Could panic the `.(string)` type conversion if `ctx.Value()` is not defined, for any reason?

Reply:
No, [type assertion](https://golang.org/ref/spec#Type_assertions) returns `false` for `nil`.